Add a tutorial on newly introduce Spell Checker

lebaudantoine opened the pull request:
To-do:
- [ ] Fix vertical overflow
- [ ] ~~Style the pop-in~~, edit: won't lose time on a tutorial seen once, 2min.
- [x] Disable interactivity of high-lighted elements

Brainsto:
- Add more steps?
- Specify a unique cookie name to prevent potential collisions when more tutorials are added in the future
- Improve the UX by making the Grist form more visible (currently, it's quite hidden).
- Have a way to trigger the tutorial manually?
- How to avoid increase in support if user are unhappy?

Commit:
➕(frontend) add intro.js dependency

Based on @manuhabitela suggestion, I chose Intro.js to build any onboarding
tutorial on Pad's new features.

Intro.js license should permit usage in non-commercial open source projects.

lebaudantoine committed:
🚧(frontend) introduce tutorial presenting spell-checker

Using Intro.js, bootstrap a basic tooltip to inform users about the newly
introduced spell-checker feature in beta.

The tutorial is essential to ensure users are aware of this new functionality
and how to enable or disable it. I've added a 'boilerplate' content, open to
suggestions and improvements.

Will be enhanced in upcoming commits.

lebaudantoine committed:
💄(frontend) make highlight border visible on dark editor

Intro.js uses a black border to highlight the focused element, which is
invisible against our editor's default black background.

Applied minor style adjustments to improve UX on this feature.

lebaudantoine committed:
🚸(frontend) avoid showing bullets for a 2-steps tutorial

Seemed overkilled to me. With only two steps, these bullets
don't provide any value.

Plus they are clickable, which is really redundant with prev
and next buttons when there are only 2 steps.

lebaudantoine committed:
🎨(frontend) hide tutorial if spell-checker is disabled

Prevents displaying tutorial steps for elements not present in the DOM
when the spell-checker is disabled.

Also improved elements' naming for better readability.

lebaudantoine committed:
🚸(frontend) show tutorial only once

I struggled using the IntroJs API and might have missed something.
There's no built-in way to force the tutorial to show only once without
displaying a checkbox to the user.

I prefer not to give users the option to skip the tutorial. It should be
shown only once.

Without using the 'dontShowAgain' option, the 'isActive' method always
returns true, even if the cookie is set to true.

To handle this, I used the 'dontShowAgain' option, set the cookie
value, and override some styling to hide the checkbox.

While developing, clear the a cookie named 'introjs-dontShowAgain'.

lebaudantoine commented:
As of tonight, the current styling of the tutorial:


https://github.com/user-attachments/assets/2b2728a1-85fc-43b1-b5d8-fcd5266b5aac

manuhabitela commented on public/js/lib/spell-checker/tutorial.js:
```diff
@@ -0,0 +1,39 @@
+        element: spellCheckToggle
+      },
+      {
+        title: 'Feedbacks',
+        intro: "Ce statut vous informera de l'état de votre document. Une fois le tutoriel terminé, cliquez dessus à tout moment pour partager vos feedbacks.",
```
My 2 cents: I'd stay in french, would not talk in future tense, and wouldn't talk about "status".

For example:

```suggestion
        title: 'Alertes et retours',
        intro: "Cette icône vous alerte en temps réel sur l'orthographe de votre document. \n\nLors de la rédaction d'un document, cliquez dessus pour nous partager vos retours sur la fonctionnalité.",
```

manuhabitela commented on public/js/lib/spell-checker/tutorial.js:
```diff
@@ -0,0 +1,39 @@
+  }).start()
+  // Set a cookie to prevent showing the tutorial again
+  // This ensures the tutorial is shown only once
+  introJs().setDontShowAgain(true)
```
How about:

- setDontShowAgain only [oncomplete](https://introjs.com/docs/tour/api#introjsoncompleteprovidedcallback) and onexit (dont know if both are required)?
- set [exitOnOverlayClick](https://introjs.com/docs/tour/options#exitonoverlayclick) to `false` so that a misclick doesn't make us miss the tutorial? It has a close button, only two steps, and we can press esc. I feel the overlay click is more misleading than helpful.

edit: tried an implementation below

manuhabitela commented:
On my firefox it renders like this on the staging platform: :grimacing: 

edit: fixed that with an example implementation below

![image](https://github.com/user-attachments/assets/149a1b4b-8be2-466c-bfda-dbe389ced23a)

manuhabitela commented:
Dont know if I understand correctly but when testing locally as a user, I had to explicitly enable the feature by clicking the check icon in the bottom so that the tutorial shows up. Is it what we really want?

I guess we could just check for the presence of the spellcheck feature flag enabled, and add a "fake" status icon just for the tutorial if the spellcheck icon is not checked by default?

manuhabitela commented:
> Disable interactivity of high-lighted elements

You can trick the introjs code by adding some CSS: 

```css

.introjs-showElement.introjs-showElement {
  z-index: 1 !important;
}
```

manuhabitela commented:
Here is my take on it:

```diff
// javascript
-  introJs().setOptions({
+  const intro = introJs().setOptions({
    dontShowAgain: true, // Override styles to hide the checkbox
    nextLabel: 'Suivant',
    prevLabel: 'Précédent',
    doneLabel: 'Terminé',
    highlightClass: 'custom-introjs-helperLayer',
    showBullets: false,
+   disableInteraction: true,
+   exitOnOverlayClick: false,
    steps: [
      {
        title: 'Nouveauté',
       intro: "La correction orthographique est disponible en version bêta. Vous pouvez l'activer ou la  désactiver en cochant cette case.",
        element: spellCheckToggle
      },
      {
        title: 'Feedbacks',
       intro: "Ce statut vous informera de l'état de votre document. Une fois le tutoriel terminé, cliquez  dessus à tout moment pour partager vos feedbacks.",
        element: spellCheckStatus
      }
    ]
-  }).start()
+  })
+  intro.onexit(() => {
+    document.querySelector('body').classList.remove('introjs-tour')
+    intro.setDontShowAgain(true)
+  })
+  intro.start()

// css
+ body.introjs-tour {
+     overflow: hidden;
+ }
```

- this makes the tutorial more user-error proof without the overlay click
- we set to not show it again after we exited (on esc key, on close icon, on tutorial steps ended), making sure just a page reload doesn't hide it
- we disable interaction on items
- the overflow hidden tricks resolves your vertical overflow bullet point that created the positioning issue of my screenshot

This doesn't solve my [previous comment](https://github.com/betagouv/pad.numerique.gouv.fr/pull/60#issuecomment-2247470978) though and I also didn't include my wording suggestions, I'll let you do what you wish.

lebaudantoine committed:
🩹(frontend) fix tutorial bulky wip code

Thx @manuhabitela for your feedbacks and suggestions.
I've integrated everything.

- disable interaction on items
- avoid closing the tutorial by error
- making sure a page reload doesn't hide it
- fix vertical overflow

(The vertical overflow still persists ...)

lebaudantoine committed:
🚸(frontend) improve feedback form visibility

Accessing the feedback form was not intuitive as it was hidden within
the spell-checker status element, causing confusion by mixing responsibilities.

Added a dedicated item to the 'statusbar' HTML element improving access
to the feedback form. Additionally, the tutorial has been updated
with an extra step to explicitly demonstrate the new feedback
loop access point.

Missing an environment variable for the form URL, but this value is
not expected to change soon.

The chosen term 'feedbacks' maintains consistency with existing English
wording in the status bar.

lebaudantoine committed:
♻️(frontend) specify a unique cookie name to the tutorial

Specify a unique cookie name to prevent potential collisions
when more tutorials are added in the future

lebaudantoine committed:
✏️(frontend) fix misc typos

sorryyyy the nit-picker is back,
fixed typos while working on the editor.js.

lebaudantoine committed:
🚚(frontend) enable by default the new spellcheck

The new spellcheck will be enabled by default, until the user
explicitly disable it by toggling the check.

Renaming the cookie will ignore previous user preferences, and is
necessary to enable the new spellcheck.

(It's kind of a bad technic to 'force' adoption, I am not proud of it...)

lebaudantoine committed:
💄(frontend) fix undesired transparency

I've no idea why this element was transparent. Fixed it!

manuhabitela committed:
💄(frontend) fix firefox bug on feedback tutorial
